File: Huff.cs
// Bradford Arrington 2025
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;

partial class Compressor
{
    const int END_OF_STREAM = 256;
    const int EOF = -1; // Define a custom EOF
	
    public struct Node
    {
        public uint Count { get; set; }
        public uint SavedCount { get; set; }
        public int Child0 { get; set; }
        public int Child1 { get; set; }
    }
    public struct Code
    {
        public ulong CodeValue;
        public int CodeBits;
    }

    public string CompressionName = "static order 0 model with Huffman coding";
    public static string Usage = "infile outfile [-d]\n\nSpecifying -d will dump the modeling data\n";

    public static void FilePrintBinary(Stream file, uint code, int bits)
    {
        uint mask = 1u << (bits - 1);
        while (mask != 0)
        {
            if ((code & mask) != 0)
                file.WriteByte((byte)'1');
            else
                file.WriteByte((byte)'0');
            mask >>= 1;
        }
    }

    public void CompressFile(FileStream input, BitFile output, int argc, string[] argv)
    {

        ulong[] counts = new ulong[256];
        for (int i = 0; i < counts.Length; i++)
        {
            counts[i] = (uint)(i + 1); // Initialize with some values
        }
		//List<ulong> counts = new List<ulong>();
		
        Node[] nodes = new Node[514];
        Code[] codes = new Code[257];
        CountBytes(input, counts);
        ScaleCounts(counts, nodes);
        OutputCounts(output, nodes);
        int rootNode = BuildTree(nodes);
        ConvertTreeToCode(nodes, codes, 0, 0, rootNode);
        if (argc > 0 && argv[0] == "-d")
            PrintModel(nodes, codes);
        CompressData(input, output, codes);
    }

    public void ExpandFile(BitFile input, FileStream output, int argc, string[] argv)
    {
        Node[] nodes = new Node[514];
        int rootNode;

        InputCounts(input, nodes);
		//PrintModel(nodes, null);
        rootNode = BuildTree(nodes);
		Console.WriteLine("model 2");
	    //PrintModel(nodes, null);
        if (argc > 0 && argv[0] == "-d")
            PrintModel(nodes, null);
        ExpandData(input, output, nodes, rootNode);
    }

    public void OutputCounts(BitFile output, Node[] nodes)
    {
        int first;
        int last= 256;
        int next = 1;
        int i;

        first = 0;
        while (first < 255 && nodes[first].Count == 0) {
          first++;
		}
	    first = 10;
		Console.WriteLine(" first {0} last {1} next {2}", first.ToString(), last.ToString(), next.ToString());
        for (; first < 256; first = next)
        {
          last = first + 1;
          for (; ; )
          {
                for (; last < 256; last++)
                {
                    if (nodes[last].Count == 0) {
                     Console.WriteLine("First");
						break;
					}
                }
                last--;
                for (next = last + 1; next < 256; next++) { 
					if (nodes[next].Count != 0) {
						Console.WriteLine("Second");
						break;
					}
				}
				if (next > 255){ 
					Console.WriteLine("break Next");
					break;
				}
				if ((next - last) > 3) {
					Console.WriteLine("break next - last");         
					break;
				}
				last = next;
          }

		  try { output.fileStream.WriteByte((byte)first); }
          catch { fatal_error("first: Error writing byte counts\n"); }
		  try { output.fileStream.WriteByte((byte)last); }
          catch { fatal_error("last: Error writing byte counts\n"); }
		  
		  //Console.WriteLine(" first {0} last {1}", first.ToString(), last.ToString());
          for (i = first; i <= last; i++)
          {
			//Console.WriteLine(" node {0} ", nodes[i].Count.ToString());
            try { output.fileStream.WriteByte((byte)nodes[i].Count); }
		    catch { fatal_error("Error writing byte counts\n"); }
          }
        }
		try { output.fileStream.WriteByte((byte)0); }
		catch { fatal_error("Line 119 Error writing byte counts\n");}
    }
    public void fatal_error(string statement)
    {
        Console.WriteLine($"\n{statement}");
        Environment.Exit(1);
    }

    public void InputCounts(BitFile input, Node[] nodes)
    {
        int first;
        int last;
        int i;
        int c;

        for (i = 0; i < 256; i++)
            nodes[i].Count = 0;
		
        if ((first = input.fileStream.ReadByte()) == EOF)
            fatal_error("first Error reading byte counts\n");

        if ((last = input.fileStream.ReadByte()) == EOF)
            fatal_error("last Error reading byte counts\n");

        for (; ; )
        {
            for (i = first; i <= last; i++)
                if ((c = input.fileStream.ReadByte()) == EOF) {
                    fatal_error("Error reading byte counts\n");
				}
                else {
                    nodes[i].Count = (uint)c;
				}
            if ((first = input.fileStream.ReadByte()) == EOF)
                fatal_error("Error reading byte counts\n");
            if (first == 0)
                break;
            if ((last = input.fileStream.ReadByte()) == EOF)
                fatal_error("Error reading byte counts\n");
        }
        nodes[END_OF_STREAM].Count = 1;
    }
    uint[] numbers = new uint[256];
    public void CountBytes(Stream input, ulong[] counts)
    {
        long inputMarker;
        int c;
		int cnt=0;

        const int ulongSize = 8;
        const int arrayLength = 256;

		inputMarker = input.Position;
		List<byte> allBytes = new List<byte>();
		// ReadByte returns the next byte as an int, or -1 at EOF.
        while ((c = input.ReadByte()) != -1)
        {
            allBytes.Add((byte)c);
			counts[c]++;  //BitConverter.ToUInt64(allBytes.ToArray(), i * ulongSize);
        }

        for (int i = 0; i < ulongsToProcess; i++)
        {
            // Convert 8 bytes into a single ulong using BitConverter.
            // It reads 8 bytes starting from the calculated index in the byte list.
            //counts[i] = BitConverter.ToUInt64(allBytes.ToArray(), i * ulongSize);
        }		
       /* while ((c = input.ReadByte()) != -1) 
		{ 
			Array.Resize(ref numbers, numbers.Length + 1);
			numbers[numbers.Length - 1] = (uint)c; 
			counts[c]++;
			cnt++;
		}*/
		for (int x = 0; x < 30; x++)
        {
             Console.WriteLine($"Byte {allBytes[x]}: {counts[x]}");
        }
	    
		Console.WriteLine("marker {0} cnt {1}", inputMarker.ToString(), cnt.ToString());

        input.Seek(inputMarker, SeekOrigin.Begin);
    }

    public void ScaleCounts(ulong[] counts, Node[] nodes)
    {
        ulong maxCount = 0;
        for (int i = 0; i < 256; i++)
            if (counts[i] > maxCount)
                maxCount = counts[i];

        if (maxCount == 0)
        {
            counts[0] = 1;
            maxCount = 1;
        }

        maxCount /= 255;
        maxCount += 1;

        for (int i = 0; i < 256; i++)
        {
            nodes[i].Count = (uint)(counts[i] / maxCount);
            if (nodes[i].Count == 0 && counts[i] != 0)
                nodes[i].Count = 1;
        }
        nodes[END_OF_STREAM].Count = 1;
    }

    public int BuildTree(Node[] nodes)
    {
        int next_free;
        int i;
        int min_1;
        int min_2;
        nodes[513].Count = 0xFFFF;

        for (next_free = END_OF_STREAM + 1; ; next_free++)
        {
            min_1 = 513;
            min_2 = 513;
            for (i = 0; i < next_free; i++)
                if (nodes[i].Count != 0)
                {
                    if (nodes[i].Count < nodes[min_1].Count)
                    {
                        min_2 = min_1;
                        min_1 = i;
                    }
                    else if (nodes[i].Count < nodes[min_2].Count)
                        min_2 = i;
                }
            if (min_2 == 513)
                break;

            nodes[next_free].Count = nodes[min_1].Count + nodes[min_2].Count;
            nodes[min_1].SavedCount = nodes[min_1].Count;
            nodes[min_1].Count = 0;
            nodes[min_2].SavedCount = nodes[min_2].Count;
            nodes[min_2].Count = 0;
            nodes[next_free].Child0 = min_1;
            nodes[next_free].Child1 = min_2;
            //Console.WriteLine($" Child0 {min_1} Child1 {min_2}");
        }

        next_free--;
        nodes[next_free].SavedCount = nodes[next_free].Count;
        return next_free;
    }

    public void ConvertTreeToCode(Node[] nodes, Code[] codes, uint codeSoFar, int bits, int node)
    {
        if (node <= END_OF_STREAM)
        {
            codes[node].CodeValue = codeSoFar;
            codes[node].CodeBits = bits;
            return;
        }
        codeSoFar <<= 1;
        bits++;
        ConvertTreeToCode(nodes, codes, codeSoFar, bits, nodes[node].Child0);
        ConvertTreeToCode(nodes, codes, codeSoFar | 1, bits, nodes[node].Child1);
    }

    public void PrintModel(Node[] nodes, Code[] codes)
    {
        for (int i = 0; i < 513; i++)
        {
            if (nodes[i].SavedCount != 0)
            {
                Console.Write("node=");
                PrintChar(i);
                Console.Write($"  count={nodes[i].SavedCount,3}  child_0=");
                PrintChar(nodes[i].Child0);
                Console.Write("  child_1=");
                PrintChar(nodes[i].Child1);
                if (codes != null && i <= END_OF_STREAM)
                {
                    Console.Write("  Huffman code=");
                    FilePrintBinary(Console.OpenStandardOutput(), (uint)codes[i].CodeValue, codes[i].CodeBits);
                }
                Console.WriteLine();
            }
        }
    }

    public static void PrintChar(int c)
    {
        if (c >= 0x20 && c < 127)
            Console.Write($"'{(char)c}'");
        else
            Console.Write($"{c,3}");
    }

    public void CompressData(Stream input, BitFile output, Code[] codes)
    {
        int c;
        while ((c = input.ReadByte()) != EOF)
            output.OutputBits((uint)codes[c].CodeValue, codes[c].CodeBits);
        output.OutputBits((uint)codes[END_OF_STREAM].CodeValue, codes[END_OF_STREAM].CodeBits);
    }

    public void ExpandData(BitFile input, Stream output, Node[] nodes, int rootNode)
    {
        int node;

        for (; ; )
        {
            node = rootNode;
            do
            {
                if (input.InputBit() == 1) {
                    node = nodes[node].Child1;
				}
                else { 
				  node = nodes[node].Child0; 
				}
			}
            while (node > END_OF_STREAM);
			
            if (node == END_OF_STREAM)
                break;
            try { 
				output.WriteByte((byte)node); 
			}
			catch {
                fatal_error("ExpandData: Error trying to write expanded byte to output");
			}
        }
		output.Close();
    }
}

File: Python/huff.py
#Brad Arrington
import sys
from collections import namedtuple

END_OF_STREAM = 256
COMPRESSION_NAME = "static order 0 model with Huffman coding"
USAGE = "infile outfile [-d]\n\nSpecifying -d will dump the modeling data\n"
    
def FilePrintBinary(file, code, bits):
    """Placeholder for printing the binary code."""
    print(f"<{code:0{bits}b}>", end="")
    pass

class Node:
    def __init__(self):
        self.count = 0
        self.saved_count = 0
        self.child_0 = 0
        self.child_1 = 0

class Code:
    def __init__(self):
        self.code = 0
        self.code_bits = 0

def LINE():
    return sys._getframe(1).f_lineno

def compress_file(input_file: FileIO, output_bit_file: 'CompressorBitio.BitFile',  argc: int, argv: list[str]):
    counts = [0] * 256
    nodes = [Node() for _ in range(514)]
    codes = [Code() for _ in range(257)]

    count_bytes(input_file, counts)
    scale_counts(counts, nodes)
    output_counts(output_bit_file, nodes)
    root_node = build_tree(nodes)
    convert_tree_to_code(nodes, codes, 0, 0, root_node)
    
    if len(argv) > 0 and argv[0] == "-d":
        print_model(nodes, codes)

    compress_data(input_file, output_bit_file, codes)

def expand_file(input_bit_file: 'CompressorBitio.BitFile', output_file: FileIO, argc: int, argv: list[str]):
    nodes = [Node() for _ in range(514)]

    input_counts(input_bit_file, nodes)
    root_node = build_tree(nodes)

    if len(argv) > 0 and argv[0] == "-d":
        print_model(nodes, None) # Pass None instead of 0 for codes

    expand_data(input_bit_file, output_file, nodes, root_node)

def output_counts(output_bit_file, nodes):
    last = 256
    next_ = 1

    first = 0
    # Find the first non-zero count node

    while first < 255 and nodes[first].count == 0:
        first += 1

    while first < 256:
        last = first + 1

        while True:
            # Find first zero-count node
            while last < 256:
                if nodes[last].count == 0:
                    break
                last += 1

            last -= 1

            # Find next nonzero-count node
            next_ = last + 1
            while next_ < 256:
                if nodes[next_].count != 0:
                     break
                next_ += 1

            if next_ > 255:
                break

            if (next_ - last) > 3:
                break

            last = next_

        try:
            output_bit_file.file_stream.write(bytes([first]))
        except Exception:
            print("Error writing byte counts (range)", LINE())
            
        try:
            output_bit_file.file_stream.write(bytes([last]))
        except Exception:
            print("Error writing byte counts (range)", LINE())
            
        for i in range(first, last + 1):
            try:
                # Assuming scaled count fits in one byte (max count is <= 255 after scaling).
                output_bit_file.file_stream.write(bytes([nodes[i].count])) # != nodes[i].count:
            except Exception:
                print("Error writing byte counts (data)",  LINE())
        
        first = next_
    
    # Write the termination marker (first == 0)
    try:
        output_bit_file.file_stream.write(bytes(1))
    except Exception:
        print("Error writing byte counts (terminator)",  LINE())


def input_counts(input_bit_file, nodes):
    for i in range(256):
        nodes[i].count = 0

    value = input_bit_file.file_stream.read(1)
    first = value[0]
    if first == -1: 
        print("Error reading byte counts (first)",  LINE())

    value = input_bit_file.file_stream.read(1)
    last = value[0]
    if last == -1: 
        print("Error reading byte counts (last)",  LINE())

    while True:
        for i in range(first, last + 1):
            value = input_bit_file.file_stream.read(1)
            c = value[0]
            if c == -1: 
                print("Error reading byte counts (data)",  LINE())
            else:
                nodes[i].count = c # c is the count (0-255)

        # Read the next 'first' marker
        value = input_bit_file.file_stream.read(1)
        first = value[0]
        if first == -1: 
            print("Error reading byte counts (next first)",  LINE())

        if first == 0:
            break # Termination marker found

        # Read the next 'last' marker
        value = input_bit_file.file_stream.read(1)
        last = value[0]
        if last == -1: 
            print("Error reading byte counts (next last)",  LINE())

    # Set EOF count
    nodes[END_OF_STREAM].count = 1


def count_bytes(input_file, counts):
    input_marker = input_file.tell()

    input_file.seek(0) # Ensure we read from the start for counting
    while True:
        c = input_file.read(1)
        if not c:
            break # EOF
        c_val = c[0]
        counts[c_val] += 1
    
    print(f"len {len(counts)}")
    #for i in range(30):
    #  print(f" counts {counts[i]}")
      
    input_file.seek(input_marker)


def scale_counts(counts, nodes):
    max_count = 0
    for count in counts:
        if count > max_count:
            max_count = count

    if max_count == 0:
        counts[0] = 1
        max_count = 1

    # C code: max_count = max_count / 255; max_count = max_count + 1;
    # This ensures the scaled counts fit within a byte (0-255)
    max_count = max_count // 255
    max_count = max_count + 1

    for i in range(256):
        scaled_count = counts[i] // max_count
        # Ensure any non-zero count is at least 1 after scaling
        if scaled_count == 0 and counts[i] != 0:
            scaled_count = 1
        nodes[i].count = scaled_count

    # Set EOF count
    nodes[END_OF_STREAM].count = 1


def build_tree(nodes):
    nodes[513].count = 0xFFFF # High value (65535)
    next_free = END_OF_STREAM + 1 # Start non-leaf nodes after 257 (0-256 for chars + EOF)

    while True:
        min_1 = 513
        min_2 = 513

        # Find two nodes with the minimum non-zero counts
        for i in range(next_free):
            if nodes[i].count != 0:
                if nodes[i].count < nodes[min_1].count:
                    min_2 = min_1
                    min_1 = i
                elif nodes[i].count < nodes[min_2].count:
                    min_2 = i

        if min_2 == 513:
            break # Tree is complete (only one node left, min_1)

        nodes[next_free].count = nodes[min_1].count + nodes[min_2].count
        
        # Save counts and zero them out to prevent re-selection
        nodes[min_1].saved_count = nodes[min_1].count
        nodes[min_1].count = 0
        nodes[min_2].saved_count = nodes[min_2].count
        nodes[min_2].count = 0

        # Set children
        nodes[next_free].child_0 = min_1
        #print(f"child0_bt {min_1}")
        nodes[next_free].child_1 = min_2
        #print(f"child1_bt {min_2}")
        
        next_free += 1

    # next_free points one past the root node, so decrement to get the root's index
    next_free -= 1
    nodes[next_free].saved_count = nodes[next_free].count # Save root count
    return next_free # Return root node index

def convert_tree_to_code(nodes, codes, code_so_far, bits, node):
    if node <= END_OF_STREAM: # Leaf node (character or EOF)
        codes[node].code = code_so_far
        codes[node].code_bits = bits
        return

    code_so_far <<= 1
    bits = bits + 1
    #print(f"child0_ttc {nodes[node].child_0}")
    convert_tree_to_code(nodes, codes, code_so_far, bits, nodes[node].child_0)
    convert_tree_to_code(nodes, codes, code_so_far | 1, bits, nodes[node].child_1)

def print_char(c):
    if 0x20 <= c < 127:
        print(f"'{chr(c)}'", end="")
    else:
        print(f"{c:3d}", end="")

def file_print_binary(file: FileIO, code: int, bits: int):
    mask = 1 << (bits - 1)
    while mask != 0:
        if code & mask:
            file.write(b'1')
        else:
            file.write(b'0')
        mask >>= 1

def print_model(nodes, codes):
    for i in range(513):
        if nodes[i].saved_count != 0:
            print(f"node=", end="")
            print_char(i)
            print(f"  count={nodes[i].saved_count:3d}", end="")
            
            print("  child_0=", end="")
            print_char(nodes[i].child_0)
            print("  child_1=", end="")
            print_char(nodes[i].child_1)

            if codes is not None and i <= END_OF_STREAM:
                print("  Huffman code=", end="")
                FilePrintBinary(sys.stdout.buffer, codes[i].code, codes[i].code_bits)
            
            print() # Newline


def compress_data(input_file, output_bit_file, codes):
    input_file.seek(0)
    
    while True:
        c = input_file.read(1)
        if not c:
            break # EOF
        c_val = c[0]
        #print(chr(c[0]))
        
        # Output Huffman code for the byte
        #print(f"{codes[c_val].code}  {codes[c_val].code_bits}")
        output_bit_file.output_bits(codes[c_val].code, codes[c_val].code_bits)

    output_bit_file.output_bits(codes[END_OF_STREAM].code, codes[END_OF_STREAM].code_bits)


def expand_data(input_bit_file, output_file, nodes, root_node):
    while True:
        node = root_node
        
        # Traverse the tree bit by bit
        while node > END_OF_STREAM:
            # InputBit is a placeholder function
            if input_bit_file.input_bit():
                node = nodes[node].child_1 # 1 bit (right)
            else:
                node = nodes[node].child_0 # 0 bit (left)

        # Leaf node found
        if node == END_OF_STREAM:
            break # Done

        try:
            output_file.write(bytes([node]))
        except Exception:
            print("Error trying to write expanded byte to output",  LINE())

